### ISSUE TITLE
fillna on dataset converts all variables to float

### ISSUE THREAD

**Issue by Hoeze** [pydata/xarray#3570]
#### MCVE Code Sample
```python
xr.Dataset(
     {
         "A": ("x", [np.nan, 2, np.nan, 0]),
         "B": ("x", [3, 4, np.nan, 1]),
         "C": ("x", [True, True, False, False]),
         "D": ("x", [np.nan, 3, np.nan, 4])
     },
     coords={"x": [0, 1, 2, 3]}
).fillna(value={"A": 0})
```
```
<xarray.Dataset>
Dimensions:  (x: 4)
Coordinates:
  * x        (x) int64 0 1 2 3
Data variables:
    A        (x) float64 0.0 2.0 0.0 0.0
    B        (x) float64 3.0 4.0 nan 1.0
    C        (x) float64 1.0 1.0 0.0 0.0
    D        (x) float64 nan 3.0 nan 4.0
```

#### Expected Output
```
<xarray.Dataset>
Dimensions:  (x: 4)
Coordinates:
  * x        (x) int64 0 1 2 3
Data variables:
    A        (x) float64 0.0 2.0 0.0 0.0
    B        (x) float64 3.0 4.0 nan 1.0
    C        (x) bool True True False False
    D        (x) float64 nan 3.0 nan 4.0
```

#### Problem Description
I'd like to use `fillna` to replace NaN's in some of a `Dataset`'s variables.
However, `fillna` unexpectably converts all variables to float, even if they are boolean or integers.

Would it be possible to apply `fillna` only on float / object types and consider the `value` argument, if I only want to apply `fillna` to a subset of the dataset?

#### Output of ``xr.show_versions()``
<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-957.27.2.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.4
libnetcdf: 4.6.1

xarray: 0.14.0
pandas: 0.25.1
numpy: 1.17.2
scipy: 1.3.1
netCDF4: 1.4.2
pydap: None
h5netcdf: 0.7.4
h5py: 2.9.0
Nio: None
zarr: 2.3.2
cftime: 1.0.3.4
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: None
dask: 2.5.2
distributed: 2.5.2
matplotlib: 3.1.1
cartopy: None
seaborn: 0.9.0
numbagg: None
setuptools: 41.4.0
pip: 19.2.3
conda: None
pytest: 5.0.1
IPython: 7.8.0
sphinx: None

</details>

**Comment by mathause** [comment]
The problem is that xarray calls `where(isnull(data), data, other)` for *all* variables. It also uses `dataset_join="left"`, thus `other` is `NaN` for all DataArrays that are not passed... I think `fillna` should be made a no-op `if isnull(other) or not mask.any()`


https://github.com/pydata/xarray/blob/66ab0ae4f3aa3c461357a5a895405e81357796b1/xarray/core/duck_array_ops.py#L284-L288

A possible workaround is to use `da.fillna(value={"A": 0, "C": True})`

**Comment by mathause** [comment]
The straightforward idea unfortunately leads to some test failures. I.e. the following does not work:

```python
def fillna(data, other):
    # we need to pass data first so pint has a chance of returning the
    # correct unit
    # TODO: revert after https://github.com/hgrecco/pint/issues/1019 is fixed
    mask = isnull(data)

    if mask.any():
        return where(~mask, data, other)
    else:
        return data
```

**Comment by keewis** [comment]
yes, we would need to `compute` if we wanted to use `any` / `all` to detect missing values. I guess we could check if the dtype allows missing values (using something like `dtype.kind in "cfO"`), and only replace for those that do.

**Comment by mathause** [comment]
There are 3 types of failures:
- non-lazy evaluation
- missing alignment (in merge)
- not raising `DimensionalityError`

So its not trivial...

<details>

```Python traceback
FAILED xarray/tests/test_dask.py::test_lazy_array_equiv_merge[no_conflicts] - RuntimeError: Too many computes. Total: 1 > max: 0.
FAILED xarray/tests/test_dataarray.py::TestReduce2D::test_idxmin[True-x2-minindex2-maxindex2-nanindex2] - RuntimeError: Too many computes. Total: 2 > max: 1.
FAILED xarray/tests/test_dataarray.py::TestReduce2D::test_idxmax[True-x2-minindex2-maxindex2-nanindex2] - RuntimeError: Too many computes. Total: 2 > max: 1.
FAILED xarray/tests/test_dataset.py::TestDataset::test_dask_is_lazy - xarray.tests.UnexpectedDataAccess: Tried accessing data
FAILED xarray/tests/test_merge.py::TestMergeMethod::test_merge_broadcast_equals - ValueError: applied function returned data with unexpected number of dimensions. Received 0 dimension(s) but expected 1 dim...
FAILED xarray/tests/test_units.py::TestDataArray::test_fillna[int64-python_scalar-no_unit] - Failed: DID NOT RAISE <class 'pint.errors.DimensionalityError'>
... more of those
```
</details>

Potentially this could also be fixed in `Dataset.fillna`, however, the fill value would need to be `dtype` dependent, so also not trivial...

https://github.com/pydata/xarray/blob/66ab0ae4f3aa3c461357a5a895405e81357796b1/xarray/core/dataset.py#L4000

**Comment by dcherian** [comment]
> I guess we could check if the dtype allows missing values (using something like dtype.kind in "cfO")

This seems sensible; `fillna` only makes sense for dtypes that allow NA.
